File: Binary Search/1011. Capacity To Ship Packages Within D Days/Solution.py
from typing import List

class Solution:
    def shipWithinDays(self, weights: List[int], days: int) -> int:
        # N : len(weights), M : sum(weights) - max(weights)
        # Time Complexity: O(M * N)
        # Space Complexity: O(1)
        # n = len(weights)
        # m = max(weights)
        # s = sum(weights) + 1
        # minWeight = float('inf')
        # for i in range(m, s):
        #     dayCount = 0
        #     flag = True
        #     summ = 0
        #     for weight in weights:
        #         summ += weight
        #         if summ > i:
        #             summ = weight
        #             dayCount += 1
        #         if dayCount >= days:
        #             flag = False
        #             break
        #     if flag:
        #         return i
        # return 0



        # N : len(weights), M : sum(weights) - max(weights)
        # Time Complexity: O(N * logM)
        # Space Complexity: O(1)
        l = max(weights)
        r = sum(weights) + 1
        while l <= r:
            mid = (r + l) // 2
            dayCount = 1
            summ = 0
            for weight in weights:
                summ += weight
                if summ > mid:
                    summ = weight
                    dayCount += 1
            if dayCount <= days:
                r = mid - 1
            else:
                l = mid + 1
        return l

File: Binary Search/1011. Capacity To Ship Packages Within D Days/Solution.java
import java.util.*;

class Solution {
    public int shipWithinDays(int[] weights, int days) {
        // N : len(weights), M : sum(weights) - max(weights)
        // Time Complexity: O(M * N)
        // Space Complexity: O(1)
        // int n = weights.length;
        // int m = Arrays.stream(weights).max().getAsInt();
        // int s = Arrays.stream(weights).sum();
        // for (int i = m; i <= s; i++) {
        //     int dayCount = 0;
        //     int totalWeight = 0;
        //     boolean flag = true;
        //     for (int weight: weights) {
        //         totalWeight += weight;
        //         if (totalWeight > i) {
        //             totalWeight = weight;
        //             dayCount++;
        //         }
        //         if (dayCount >= days) {
        //             flag = false;
        //             break;
        //         }
        //     }
        //     if (flag) return i;
        // }
        // return 0;



        // N : len(weights), M : sum(weights) - max(weights)
        // Time Complexity: O(N * logM)
        // Space Complexity: O(1)
        int l = Arrays.stream(weights).max().getAsInt();
        int r = Arrays.stream(weights).sum();
        while (l <= r) {
            int mid = (r + l) / 2;
            int dayCount = 1;
            int totalWeight = 0;
            for (int weight: weights) {
                totalWeight += weight;
                if (totalWeight > mid) {
                    totalWeight = weight;
                    dayCount++;
                }
            }
            if (dayCount <= days) r = mid - 1;
            else l = mid + 1;
        }
        return l;
    }
}